storage/src/mmr/mem.rs

```diff
@@ -90,31 +87,51 @@ impl<H: CHasher> Mmr<H> {
     }
 
     /// Add an element to the MMR and return its position in the MMR.
+    ///
+    /// The implementation initializes a new Hasher object with each call. For bulk additions,
+    /// consider using `add_with_hasher` to avoid this overhead.
     pub fn add(&mut self, element: &H::Digest) -> u64 {
+        self.add_with_hasher(element, &mut H::new())
+    }
+
+    /// Add an element to the MMR and return its position in the MMR.
+    ///
+    /// This version requires the caller to provide an already initialized hasher for reuse to
+    /// optimize bulk additions.
+    pub fn add_with_hasher(&mut self, element: &H::Digest, hasher: &mut H) -> u64 {
         let peaks = nodes_needing_parents(self.peak_iterator());
         let element_pos = self.index_to_pos(self.nodes.len());
 
         // Insert the element into the MMR as a leaf.
-        let mut hash = Hasher::new(&mut self.hasher).leaf_hash(element_pos, element);
+        let mut hash = Hasher::new(hasher).leaf_hash(element_pos, element);
         self.nodes.push(hash.clone());
 
         // Compute the new parent nodes if any, and insert them into the MMR.
+        let mut h = Hasher::new(hasher);
```

Comment:
nit: does it make sense to reuse the hasher you made a few lines ago?

Reply:
yup fixed